3DCalendar/CalendarViewController.swift

```diff
@@ -0,0 +1,197 @@
+//
+//  CalendarViewController.swift
+//  bolt
+//
+//  Created by Najia Haider on 2/20/19.
+//  Copyright © 2019 Najia Haider. All rights reserved.
+//
+
+import UIKit
+//Calendar View Controller based off of https://github.com/jahid-hasan-polash/Calendar-iOS
+class CalendarViewController: UIViewController, UICollectionViewDelegateFlowLayout, UICollectionViewDataSource, UICollectionViewDelegate {
+ 
+    var numDaysPerMonth = [31,28,31,30,31,30,31,31,30,31,30,31]
+    var currentMonthIndex: Int = 0
+    var currentYear: Int = 0
+    var presentMonthIndex = 0
+    var presentYear = 0
+    var todaysDate = 0
+    var firstWeekDayOfMonth = 0
+    var weekdayData = ["Sun", "Mon", "Tues", "Wed", "Thurs", "Fri", "Sat"]
+    let label = UILabel(frame: CGRect(x: 0, y: 0, width: 20, height: 20))
+    var inputIndex: Int!
+    let maroon = UIColor(red: 102, green: 26, blue: 77, alpha: 1.0)
+    let greeny = UIColor(red: 51, green: 167, blue: 144, alpha: 1.0)
+    let aqua = UIColor(red: 123, green: 229, blue: 253, alpha: 1.0)
+    let lightPink = UIColor(red: 251, green: 209, blue: 223, alpha: 1.0)
+    
+    @IBOutlet private weak var collectionView: UICollectionView!
+    @IBOutlet weak var monthLabel: UILabel!
+    
+    static func make(inputIndex: Int) -> CalendarViewController {
```

Comment:
I think we can remove almost all the logic in this view controller by doing something like this:

First, we inject a date. If you remember our conversation about iterating over every day in the year using this:

```
let components = Calendar.current.dateComponents([.year], from: Date())
let startOfYear = Calendar.current.date(from: components)
```
and this:
```
let components = Calendar.current.dateComponents([.year], from: Date())
let startOfYear = Calendar.current.date(from: components)!

let monthComponent = DateComponents(month: 1)
let feb = Calendar.current.date(byAdding: monthComponent, to: startOfYear)
```

We can take all 12 of those and pass them into the `make` function of this view controller

```
static func make(inputIndex: Date) -> CalendarViewController
```

Now inside this view controller, we can remove all of the `presentYear`, `todaysDate`, `numDaysPerMonth`, `currentYear` etc stuff in favor of generating each date based on the `Date` that was passed in:

```
let datePassedIn = calendar.date(from: dateComponents)!

let range = Calendar.current.range(of: .day, in: .month, for: datePassedIn)!
let currentDate = Calendar.current.dateComponents([.year, .month],
                                                   from: Date())
let dates = range.compactMap {
    return DateComponents(calendar: .current,
                          timeZone: .current,
                          year: currentDate.year,
                          month: currentDate.month,
                          day: $0).date
}
```
Just with those few lines of code we'll have every single day for that particular month. We can use each of those days to create our table view cells. We don't have to worry about leap years or any of that stuff. Let's chat more if you have questions!

Reply:
Yeah I think we should talk in person about this because a lot of this stuff is incorporating the code from the tutorial I did!